### PR TITLE
Add ch.imatt.ClipQR

### PR DESCRIPTION
**Pull request by MatthieuBarthel**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*  
  **Note: I think `--device=all` is required for camera and screen capture access**
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by MatthieuBarthel** [comment]
bot, build ch.imatt.ClipQR

File: ch.imatt.ClipQR.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  <screenshots>
+    <screenshot type="default">
+      <image>https://gitlab.com/imatt-foss/clipqr/-/raw/master/screenshot.png</image>
+    </screenshot>
+  </screenshots>
+
+  <url type="homepage">https://gitlab.com/imatt-foss/clipqr</url>
+  <url type="bugtracker">https://gitlab.com/imatt-foss/clipqr/issues</url>
+
+  <content_rating type="oars-1.1" />
```
Comment (hfiguiere) [inline ch.imatt.ClipQR.appdata.xml:31]: missing `releases`

File: ch.imatt.ClipQR.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+command: clipqr
+
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --device=all
+
+build-options:
+  build-args: [ "--share=network" ]
```
Comment (hfiguiere) [inline ch.imatt.ClipQR.yml:16]: this will fail the build. there is no network.

File: ch.imatt.ClipQR.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  - name: clipqr
+    buildsystem: simple
+    build-commands:
+      - $GOROOT/bin/go build -o clipqr
+      - install -Dm00755 clipqr $FLATPAK_DEST/bin/clipqr
+      - >
+        for size in 16 24 32 48 64 128 256 512; do
+          rsvg-convert -w $size -h $size -f png -o $size.png icon.svg
+          install -Dm644 $size.png /app/share/icons/hicolor/${size}x${size}/apps/$FLTPAK_ID.png
+        done
```
Comment (hfiguiere) [inline ch.imatt.ClipQR.yml:31]:
```suggestion
```

You don't need this. SVG is enough.

File: ch.imatt.ClipQR.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      - install -Dm00644 $FLATPAK_ID.desktop $FLATPAK_DEST/share/applications/$FLATPAK_ID.desktop
+      - install -Dm00644 $FLATPAK_ID.appdata.xml $FLATPAK_DEST/share/appdata/$FLATPAK_ID.appdata.xml
+    sources:
+      - type: archive
+        url: https://gitlab.com/imatt-foss/clipqr/-/archive/v1.0.0/clipqr-v1.0.0.tar.gz
+        sha256: 1eb9a6ee7e431c252e30a087d0c72868c27931a62ff3219f597d78c7f3b45d7f
+      - type: file
+        path: ch.imatt.ClipQR.appdata.xml
+      - type: file
+        path: ch.imatt.ClipQR.desktop
```
Comment (hfiguiere) [inline ch.imatt.ClipQR.yml:42]: have the .desktop and appdata been submitted upstream?

A Linux desktop app without a `.desktop` is a bug.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
see https://github.com/flatpak/flatpak-builder-tools for your non vendored go dependencies.

File: ch.imatt.ClipQR.yml (hunk 2 above)
Comment (MatthieuBarthel) [reply ch.imatt.ClipQR.yml:16]: Ok many thanks for your help, I will try without this :)

File: ch.imatt.ClipQR.yml (hunk 4 above)
Comment (MatthieuBarthel) [reply ch.imatt.ClipQR.yml:37]: Nope but I will commit them upstream, thanks again

**Comment by nedrichards** [comment]
bot, build ch.imatt.ClipQR

**Comment by hfiguiere** [comment]
there is still `build-args`, so...

**Comment by MatthieuBarthel** [comment]
Yes for the moment I'm stucked with the go dependencies, I created successfully the source list but I can't build the app without network. At first I had an error because it wants to sync the vendor directory :

 <details><summary>Logs</summary>

```
Running: $GOROOT/bin/go build -mod vendor -o clipqr
go: inconsistent vendoring in /run/build/clipqr:
	fyne.io/fyne/v2@v2.1.4: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/kbinani/screenshot@v0.0.0-20210720154843-7d3a670d8329: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/makiuchi-d/gozxing@v0.1.1: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
....
	To ignore the vendor directory, use -mod=readonly or -mod=mod.
	To sync the vendor directory, run:
		go mod vendor
```
</details>

So I added  `go mod vendor` before the build command, but it still tries to connect to internet and fails:

 <details><summary>Logs</summary>

```
Running: $GOROOT/bin/go mod vendor
go: downloading github.com/kbinani/screenshot v0.0.0-20210720154843-7d3a670d8329
go: downloading github.com/pion/mediadevices v0.3.3
go: downloading github.com/makiuchi-d/gozxing v0.1.1
go: downloading fyne.io/fyne/v2 v2.1.4
go: downloading golang.design/x/clipboard v0.6.0
go: downloading golang.org/x/image v0.0.0-20220321031419-a8550c1d254a
imatt.ch/clipqr imports
	fyne.io/fyne/v2: fyne.io/fyne/v2@v2.1.4: Get "https://proxy.golang.org/fyne.io/fyne/v2/@v/v2.1.4.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	fyne.io/fyne/v2/app: fyne.io/fyne/v2@v2.1.4: Get "https://proxy.golang.org/fyne.io/fyne/v2/@v/v2.1.4.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	fyne.io/fyne/v2/canvas: fyne.io/fyne/v2@v2.1.4: Get "https://proxy.golang.org/fyne.io/fyne/v2/@v/v2.1.4.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	fyne.io/fyne/v2/container: fyne.io/fyne/v2@v2.1.4: Get "https://proxy.golang.org/fyne.io/fyne/v2/@v/v2.1.4.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	fyne.io/fyne/v2/theme: fyne.io/fyne/v2@v2.1.4: Get "https://proxy.golang.org/fyne.io/fyne/v2/@v/v2.1.4.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	fyne.io/fyne/v2/widget: fyne.io/fyne/v2@v2.1.4: Get "https://proxy.golang.org/fyne.io/fyne/v2/@v/v2.1.4.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	github.com/kbinani/screenshot: github.com/kbinani/screenshot@v0.0.0-20210720154843-7d3a670d8329: Get "https://proxy.golang.org/github.com/kbinani/screenshot/@v/v0.0.0-20210720154843-7d3a670d8329.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	github.com/makiuchi-d/gozxing: github.com/makiuchi-d/gozxing@v0.1.1: Get "https://proxy.golang.org/github.com/makiuchi-d/gozxing/@v/v0.1.1.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	github.com/makiuchi-d/gozxing/qrcode: github.com/makiuchi-d/gozxing@v0.1.1: Get "https://proxy.golang.org/github.com/makiuchi-d/gozxing/@v/v0.1.1.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	github.com/pion/mediadevices: github.com/pion/mediadevices@v0.3.3: Get "https://proxy.golang.org/github.com/pion/mediadevices/@v/v0.3.3.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	github.com/pion/mediadevices/pkg/driver/camera: github.com/pion/mediadevices@v0.3.3: Get "https://proxy.golang.org/github.com/pion/mediadevices/@v/v0.3.3.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
imatt.ch/clipqr imports
	github.com/pion/mediadevices/pkg/prop: github.com/pion/mediadevices@v0.3.3: Get "https://proxy.golang.org/github.com/pion/mediadevices/@v/v0.3.3.zip": dial tcp: lookup proxy.golang.org: Temporary failure in name resolution
...
```
</details>

Here is the build file if someone can help me:

 <details><summary>ch.imatt.ClipQR.yml</summary>

```
app-id: ch.imatt.ClipQR
runtime: org.freedesktop.Platform
runtime-version: '21.08'
sdk: org.freedesktop.Sdk
sdk-extensions:
  - org.freedesktop.Sdk.Extension.golang

finish-args:
  - --share=ipc
  - --socket=x11
  - --device=all

build-options:
  append-path: /usr/lib/sdk/golang/bin
  env:
    - GOBIN=/app/bin

modules:
  - name: clipqr
    buildsystem: simple
    build-commands:
      - go mod vendor
      - go build -o clipqr
      - install -Dm00755 clipqr $FLATPAK_DEST/bin/clipqr
      - install -Dm00644 icon.svg $FLATPAK_DEST/share/icons/hicolor/scalable/apps/$FLATPAK_ID.svg
      - install -Dm00644 $FLATPAK_ID.desktop $FLATPAK_DEST/share/applications/$FLATPAK_ID.desktop
      - install -Dm00644 $FLATPAK_ID.appdata.xml $FLATPAK_DEST/share/appdata/$FLATPAK_ID.appdata.xml
    sources:
      - type: archive
        url: https://gitlab.com/imatt-foss/clipqr/-/archive/v1.0.0/clipqr-v1.0.0.tar.gz
        sha256: 1eb9a6ee7e431c252e30a087d0c72868c27931a62ff3219f597d78c7f3b45d7f
      - type: file
        path: ch.imatt.ClipQR.appdata.xml
      - type: file
        path: ch.imatt.ClipQR.desktop
      - type: git
        url: https://github.com/fyne-io/fyne
        tag: v2.1.4
        dest: vendor/fyne.io/fyne/v2
      - type: git
        url: https://github.com/blackjack/webcam.git
        dest: vendor/github.com/blackjack/webcam
        commit: 10ed912a8539
      - type: git
        url: https://github.com/davecgh/go-spew.git
        tag: v1.1.1
        dest: vendor/github.com/davecgh/go-spew
      - type: git
        url: https://github.com/fredbi/uri.git
        dest: vendor/github.com/fredbi/uri
        commit: 3dcfdacbaaf3
      - type: git
        url: https://github.com/fsnotify/fsnotify.git
        tag: v1.5.1
        dest: vendor/github.com/fsnotify/fsnotify
      - type: git
        url: https://github.com/gen2brain/shm.git
        dest: vendor/github.com/gen2brain/shm
        commit: '083dbc7d9d83'
      - type: git
        url: https://github.com/go-gl/gl.git
        dest: vendor/github.com/go-gl/gl
        commit: 726fda9656d6
      - type: git
        url: https://github.com/go-gl/glfw.git
        dest: vendor/github.com/go-gl/glfw/v3.3/glfw
        commit: 277f93cfa958
      - type: git
        url: https://github.com/godbus/dbus.git
        tag: v5.1.0
        dest: vendor/github.com/godbus/dbus/v5
      - type: git
        url: https://github.com/goki/freetype.git
        dest: vendor/github.com/goki/freetype
        commit: 7a161fd3728c
      - type: git
        url: https://github.com/google/uuid.git
        tag: v1.3.0
        dest: vendor/github.com/google/uuid
      - type: git
        url: https://github.com/jezek/xgb.git
        tag: v1.0.0
        dest: vendor/github.com/jezek/xgb
      - type: git
        url: https://github.com/kbinani/screenshot.git
        dest: vendor/github.com/kbinani/screenshot
        commit: 7d3a670d8329
      - type: git
        url: https://github.com/lxn/win.git
        dest: vendor/github.com/lxn/win
        commit: a377121e959e
      - type: git
        url: https://github.com/makiuchi-d/gozxing.git
        tag: v0.1.1
        dest: vendor/github.com/makiuchi-d/gozxing
      - type: git
        url: https://github.com/pion/datachannel.git
        tag: v1.5.2
        dest: vendor/github.com/pion/datachannel
      - type: git
        url: https://github.com/pion/dtls.git
        tag: v2.1.3
        dest: vendor/github.com/pion/dtls/v2
      - type: git
        url: https://github.com/pion/ice.git
        tag: v2.2.2
        dest: vendor/github.com/pion/ice/v2
      - type: git
        url: https://github.com/pion/interceptor.git
        tag: v0.1.10
        dest: vendor/github.com/pion/interceptor
      - type: git
        url: https://github.com/pion/logging.git
        tag: v0.2.2
        dest: vendor/github.com/pion/logging
      - type: git
        url: https://github.com/pion/mdns.git
        tag: v0.0.5
        dest: vendor/github.com/pion/mdns
      - type: git
        url: https://github.com/pion/mediadevices.git
        tag: v0.3.3
        dest: vendor/github.com/pion/mediadevices
      - type: git
        url: https://github.com/pion/randutil.git
        tag: v0.1.0
        dest: vendor/github.com/pion/randutil
      - type: git
        url: https://github.com/pion/rtcp.git
        tag: v1.2.9
        dest: vendor/github.com/pion/rtcp
      - type: git
        url: https://github.com/pion/rtp.git
        tag: v1.7.9
        dest: vendor/github.com/pion/rtp
      - type: git
        url: https://github.com/pion/sctp.git
        tag: v1.8.2
        dest: vendor/github.com/pion/sctp
      - type: git
        url: https://github.com/pion/sdp.git
        tag: v3.0.4
        dest: vendor/github.com/pion/sdp/v3
      - type: git
        url: https://github.com/pion/srtp.git
        tag: v2.0.5
        dest: vendor/github.com/pion/srtp/v2
      - type: git
        url: https://github.com/pion/stun.git
        tag: v0.3.5
        dest: vendor/github.com/pion/stun
      - type: git
        url: https://github.com/pion/transport.git
        tag: v0.13.0
        dest: vendor/github.com/pion/transport
      - type: git
        url: https://github.com/pion/turn.git
        tag: v2.0.8
        dest: vendor/github.com/pion/turn/v2
      - type: git
        url: https://github.com/pion/udp.git
        tag: v0.1.1
        dest: vendor/github.com/pion/udp
      - type: git
        url: https://github.com/pion/webrtc.git
        tag: v3.1.27
        dest: vendor/github.com/pion/webrtc/v3
      - type: git
        url: https://github.com/pmezard/go-difflib.git
        tag: v1.0.0
        dest: vendor/github.com/pmezard/go-difflib
      - type: git
        url: https://github.com/srwiley/oksvg.git
        dest: vendor/github.com/srwiley/oksvg
        commit: 1f435e4c2b44
      - type: git
        url: https://github.com/srwiley/rasterx.git
        dest: vendor/github.com/srwiley/rasterx
        commit: 2efea2b9ea41
      - type: git
        url: https://github.com/stretchr/testify.git
        tag: v1.7.1
        dest: vendor/github.com/stretchr/testify
      - type: git
        url: https://github.com/yuin/goldmark.git
        tag: v1.4.11
        dest: vendor/github.com/yuin/goldmark
      - type: git
        url: https://github.com/golang-design/clipboard.git
        tag: v0.6.0
        dest: vendor/golang.design/x/clipboard
      - type: git
        url: https://go.googlesource.com/crypto
        dest: vendor/golang.org/x/crypto
        commit: 2c7772ba3064
      - type: git
        url: https://go.googlesource.com/exp
        dest: vendor/golang.org/x/exp
        commit: ec7cb31e5a56
      - type: git
        url: https://go.googlesource.com/image
        dest: vendor/golang.org/x/image
        commit: a8550c1d254a
      - type: git
        url: https://go.googlesource.com/mobile
        dest: vendor/golang.org/x/mobile
        commit: 34ab1303b554
      - type: git
        url: https://go.googlesource.com/net
        dest: vendor/golang.org/x/net
        commit: 27dd8689420f
      - type: git
        url: https://go.googlesource.com/sys
        dest: vendor/golang.org/x/sys
        commit: 530d0810a4d0
      - type: git
        url: https://go.googlesource.com/text
        tag: v0.3.7
        dest: vendor/golang.org/x/text
      - type: git
        url: https://go.googlesource.com/xerrors
        dest: vendor/golang.org/x/xerrors
        commit: 5ec99f83aff1
      - type: git
        url: https://gopkg.in/yaml.v3
        dest: vendor/gopkg.in/yaml.v3
        commit: 496545a6307b
```

</details>

Thanks :)

**Comment by hfiguiere** [comment]
that's because `go vendor` needs access to the network but you run it in the build phase, which is about the same as not running it at all.

See https://github.com/flatpak/flatpak-builder-tools/tree/master/go-get for helper tools.

**Comment by MatthieuBarthel** [comment]
I used flatpak-builder-tools (flatpak-go-vendor-generator.py) to build the dependencies source list, but when running flatpak-builder it miss vendor/modules.txt, that's why I believe I need to call  `go mod vendor`. 

I could manually patch modules.txt, but I think I will just commit the vendor directory in the upstream repo, it will be easier to maintain.

**Comment by MatthieuBarthel** [comment]
bot, build ch.imatt.ClipQR

**Comment by MatthieuBarthel** [comment]
LGTM

File: ch.imatt.ClipQR.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+runtime-version: '21.08'
+sdk: org.freedesktop.Sdk
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.golang
+
+command: clipqr
+
+finish-args:
+  - --share=ipc
+  - --socket=x11
```
Comment (hfiguiere) [inline ch.imatt.ClipQR.yml:12]: no wayland support?

Comment (MatthieuBarthel) [reply ch.imatt.ClipQR.yml:12]: Good question :) I will install a wayland wm and try if that works

Comment (hfiguiere) [reply ch.imatt.ClipQR.yml:12]: it's easier if you tell which toolkit it is using.

Comment (MatthieuBarthel) [reply ch.imatt.ClipQR.yml:12]: fyne.io for the GUI, wayland suport seems already in good shape: https://github.com/fyne-io/fyne/issues/1851

Comment (MatthieuBarthel) [reply ch.imatt.ClipQR.yml:12]: The app runs fine in wayland, but screen scan function doesn't work, I will try to fix this later

Comment (MatthieuBarthel) [reply ch.imatt.ClipQR.yml:12]: I took a look at this issue and it won't be easy for me to fix it, I think we can go ahead with x11, wayland support will be for another release

**Pull request closed without merging**